system/datatypes/mailinglist.php

```diff
@@ -1,406 +1,433 @@
+
+		// process form if there are no errors
+		if (!empty($form->errors)) {
+			return $this->indexViewRender($request, $form);
+		}
+
+		// check if email is already in the list, no need to have it twice.
+		$this->xml->lockAndReload();
+		$email = strtolower($form->dataFor(self::FIELD_NAME_EMAIL));
+		/** @var HyphaDomElement $addresses */
+		$addresses = $this->xml->find(self::FIELD_NAME_ADDRESSES_CONTAINER);
+		/** @var NodeList $addressCollection */
+		$xpath = '//' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_EMAIL . '="' . $email . '"]';
+		$addressCollection = $addresses->children()->findXPath($xpath);
+		$address = $addressCollection->first();
+
+		$addAddress = !$address instanceof HyphaDomElement;
+		if ($addAddress) {
+			$sendEmail = true;
+			$code = $this->constructCode();
+		} else {
+			// if the email is there just send another email if it is still pending
+			$sendEmail = $address->getAttribute(self::FIELD_NAME_STATUS) == self::ADDRESS_STATUS_PENDING;
+			$code = $address->getAttribute(self::FIELD_NAME_CONFIRM_CODE);
+		}
+
+		if ($addAddress) {
+			// add email with status pending
+			/** @var HyphaDomElement $address */
+			$address = $this->xml->createElement(self::FIELD_NAME_ADDRESS);
+			$address->setAttribute(self::FIELD_NAME_EMAIL, $email);
+			$address->setAttribute(self::FIELD_NAME_STATUS, self::ADDRESS_STATUS_PENDING);
+			$address->setAttribute(self::FIELD_NAME_CONFIRM_CODE, $code);
+			$addresses->append($address);
+			$this->xml->saveAndUnlock();
+		} else {
+			$this->xml->unlock();
+		}
+
+		if ($sendEmail) {
+			// send email so that pending subscribed can confirm email address
+			$confirmPath = str_replace('{code}', $code, self::PATH_CONFIRM_CODE);
+			$confirmUrl = $this->constructFullPath($this->pagename . '/' . $confirmPath);
+			$confirmLink = '<a href="' . $confirmUrl . '">' . __('ml-please-confirm-email') . '</a>';
+			$welcomeText = $this->xml->find(self::FIELD_NAME_EMAIL_WELCOME_TEXT)->getHtml();
+			$welcomeText .= '<br><br>' . $confirmLink;
+			$subject = hypha_getTitle() . ' - ' . __('ml-confirmation-email-subject');
+			$this->sendMail($subject, $welcomeText, [$email]);
+
+			notify('success', ucfirst(__('ml-confirmation-mail-sent')));
+		} else {
+			notify('success', ucfirst(__('ml-successfully-subscribed')));
+		}
+
+		// all is success refresh page with success notification
+		return 'reload';
+	}
+
+	private function confirmEmailAction(HyphaRequest $request) {
 		$code = isset($_GET['code']) ? $_GET['code'] : null;
 		if (null == $code) {
 			notify('error', __('missing-arguments'));
-
 			return null;
 		}
 
 		$this->xml->lockAndReload();
 
 		/** @var NodeList $addressCollection */
-		$addressCollection = $this->getDoc()->findXPath('//address[@confirm-code="' . $code . '"]');
+		$xpath = '//' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_CONFIRM_CODE . '="' . $code . '" and @' . self::FIELD_NAME_STATUS . ' != "' . self::ADDRESS_STATUS_UNSUBSCRIBED . '"]';
```

Comment:
this should also encode xpath

Reply:
done

Reply:
It breaks if I `xpath_encode` the `$code`.